sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -882,7 +882,7 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan]
     }
   }
 
-  private def inferFilters(plan: LogicalPlan): LogicalPlan = plan transform {
+  private def inferFilters(plan: LogicalPlan): LogicalPlan = plan transformUp {
```

Comment:
This fixed the SQLQueryTestSuite failure: https://github.com/apache/spark/pull/29650/checks?check_run_id=1071934445 
It is a very complex query, I'm still figuring out how to cover it in a unit test.
Other failures are covered by the "Infer all constraints from a chain of filters" UT.

Reply:
This might have been a wrong fix.

I now tried adding `right.constraints` to `LeftSemi` joins `validConstraints`. This also fixes that SQLQueryTestSuite failure. `LeftSemi` join is quite like `Inner` join, so it makes sense to have the same constraints, but I'm not 100% sure how having constraints referencing missing columns affects everything. The `InferFiltersFromConstraints.inferNewFilter` will remove those in the end.

Reply:
I believe that the UT "Infer from right side of left semi join" now covers this